[layout-test-worker-parser] Doesn't work with S3 logs.

bls1999 opened the pull request:
#### 68e07e8029ac0f9a39a10d02cfafdd0584594f17
<pre>
[layout-test-worker-parser] Doesn&apos;t work with S3 logs.
<a href="https://bugs.webkit.org/show_bug.cgi?id=288434">https://bugs.webkit.org/show_bug.cgi?id=288434</a>
<a href="https://rdar.apple.com/145524609">rdar://145524609</a>

Reviewed by Jonathan Bedard.

The layout-test-worker-parser script doesn&apos;t work with S3 logs. In practice,
this means that layout test runs that have their logs uploaded to S3 won&apos;t
parse the desired logs.

This change adds support for runs with layout test logs in S3, as well as runs
with multiple steps containing layout test results.

* Tools/Scripts/layout-test-worker-parser:
    -&gt; (load_from_url): Renamed from load_from_buildbot, add handling for S3 log links.
    -&gt; (main): Add new parameters to load_from_url function.
    -&gt; (init): Add new argument --buildbot-step.
    -&gt; (parse_raw_stdio_into_test_names): Bail if no tests found in data.
    -&gt; (parse_raw_stdio_into_workers): Ditto.

Canonical link: <a href="https://commits.webkit.org/302679@main">https://commits.webkit.org/302679@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/94a0eae078ffd51f084e3c446f5d9291ee29d803

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129897 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/2158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40755 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/137290 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/81386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131768 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2111 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/2049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/137290 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/81386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132844 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/1591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/116338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/137290 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/1502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/34467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/80560 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/110009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/34972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/139771 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/1952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/2049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/139771 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/1997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/112686 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/139771 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/31159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/54754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20260 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/2025 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/1839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/1874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/1948 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/layout-test-worker-parser:
```diff
@@ -237,22 +251,74 @@ def load_from_file(path):
+    info('Loading from URL (this may take a moment)...')
+
+    try:
+        data = requests.get(base_url)
```
Why are we loading the `base_url` instead of the `uobj` you just created? `base_url` usually means you're mutating the URL into some final form. Seems like `base_url` is really `constructed_url`, or something to that effect.

bls1999 replied:
Good point, I fixed this in the latest commit

JonWBedard commented on Tools/Scripts/layout-test-worker-parser:
```diff
@@ -237,22 +251,74 @@ def load_from_file(path):
+    layout_test_step = {}
+    build_data = json.loads(data.text)
+    for step in build_data['steps']:
+        if step['name'] == 'layout-test':
```
Seems like this should be a regex rather than a bunch of `if` statements.

bls1999 replied:
I'm not sure a regex is necessary here. Since each of the `if` statements check distinct, non-abstract conditions, a regex seems like overkill.

JonWBedard commented on Tools/Scripts/layout-test-worker-parser:
```diff
@@ -237,22 +251,74 @@ def load_from_file(path):
+
+    cstep_names = list(candidate_steps.keys())
+    if len(cstep_names) == 0:
+        # Preserve functionality for build.webkit.org runs pre-S3
```
Just note that not all buildbot instances upload all logs to S3

bls1999 committed:
[layout-test-worker-parser] Doesn't work with S3 logs.
https://bugs.webkit.org/show_bug.cgi?id=288434
rdar://145524609

Reviewed by Jonathan Bedard.

The layout-test-worker-parser script doesn't work with S3 logs. In practice,
this means that layout test runs that have their logs uploaded to S3 won't
parse the desired logs.

This change adds support for runs with layout test logs in S3, as well as runs
with multiple steps containing layout test results.

* Tools/Scripts/layout-test-worker-parser:
    -> (load_from_url): Renamed from load_from_buildbot, add handling for S3 log links.
    -> (main): Add new parameters to load_from_url function.
    -> (init): Add new argument --buildbot-step.
    -> (parse_raw_stdio_into_test_names): Bail if no tests found in data.
    -> (parse_raw_stdio_into_workers): Ditto.

Canonical link: https://commits.webkit.org/302679@main
